R/functions.R

```diff
@@ -343,3 +366,165 @@ ate = function(Z,Y){
+#' @export
+sample_mod = function(k){
+  sample(1:k,1)
+}
+
+
+#' this function gives the exposure statuses for all observations
+#'
+#' @param housestruct The house structure vector that gives the number of kids in each household
+#' @param K1 The number of houses to treat. 
+#'
+#' @return A list specifies treated houses and treated kids
+#' @export
+out_treat_household = function(housestruct, K1){
+  which_house = sample(1:length(housestruct),size=K1,replace=F) # which households to treat?
+  which_kid = apply(as.matrix(housestruct[which_house]),1,sample_mod) # which kids to treat (conditional on the treated households)?
+  return(list(which_house=which_house,which_kid=which_kid))
+}
+
+#' Gives a range of indexes that are exposed to treatments
+#'
+#' @param ii the index of our interest
+#' @param lind the first index in each household
+#' @param hind the last index in each household
+#'
+#' @return A vector of index exposed to the treament
+#' @export
+out_exp_ind = function(ii,lind,hind){
+  lb = lind[ii]
+  ub = hind[ii]
+  lb:ub
+}
+
+#' This function returns the treatment status for all observations. 
+#'
+#' @param N The number of observations (kids).
+#' @param K The number of houses.
+#' @param treatment_list A list of treated houses and treated kids.
+#' @param housestruct House structure vector that gives number of kids in each household.
+#'
+#' @return an N x 3 matrix (N = # of observations (kids)), the first column identifies which households are treated with a "1" for the first kid in a treated household, the second and third columns are the ones that matter, they are the treatment and exposure vectors at the observation (kid) level.
+#' @export
+out_Z_household = function(N,K,treatment_list,housestruct){
+  which_house = sort(treatment_list$which_house)
+  theord = order(treatment_list$which_house)
+  which_kid = treatment_list$which_kid[theord]
+  
+  Z = matrix(0,nrow=N,ncol=3)
+  colnames(Z) = c('houseind','treat','exp')
+  houseind = cumsum(housestruct)-housestruct[1]+1
+  Z[houseind,1] = 1 # separates out the different houses
+  treatedind = houseind[which_house] + which_kid - 1
+  Z[treatedind,2] = 1
+  lind = houseind[which_house] 
+  hind = lind + housestruct[which_house] - 1
+  expind = c(apply(as.matrix(1:length(which_house)),1,out_exp_ind,lind=lind,hind=hind))
+  Z[expind,3] = 1
+  Z
+}
+
+#' Returns matrix \code{Z} of dimension (number of units x number of randomizations, i.e. assignments.)
+#'
+#' @param N The number of observations (kids).
+#' @param K The number of houses.
+#' @param equal A binary parameter that determines the exact house structure. If equal=TRUE, every household has equal number of kids. If equal=FALSE, every household samples the number of kids at random.
+#' @param numrand The number of randomizations
+#' 
+#' @return A matrix compounds treated units as well as exposure functions. 0 represents a pure control unit. 1 represents a spillover unit. And 2 represents a treated unit.
```

Comment:
"A matrix of dimension (number of units x number of randomizations, i.e. assignments.) comprised of 0, denoting a unit is pure control under that particular assignment, 1 denoting a unit is a spillover under that particular assignment, and 2, denoting a unit is treated under that particular assignment."